{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e80e1158",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gc \n",
    "import os\n",
    "import sys\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from datetime import datetime\n",
    "import importlib\n",
    "\n",
    "import celloracle as co\n",
    "co.__version__\n",
    "\n",
    "sys.path.insert(0, \"/home/michal.kubacki/Githubs/Re-MEND/code/External_Datasets/GeneSet_Derivation/Herring_celloracle/helpers\")\n",
    "\n",
    "\n",
    "import config\n",
    "importlib.reload(config)\n",
    "from config import *\n",
    "n_cpus = 32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5224c926",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = \"hg19\"\n",
    "\n",
    "neurons_set = \"all_excitatory_all_ages\"\n",
    "# neurons_set = \"all_excitatory\"\n",
    "\n",
    "cells_dict = {\n",
    "    \"all_excitatory\"            :   ['L5-6_TLE4', 'L2-3_CUX2', 'L4_RORB', 'L5-6_THEMIS', 'PN_dev'],\n",
    "    \"all_excitatory_all_ages\"   :   ['L5-6_TLE4', 'L2-3_CUX2', 'L4_RORB', 'L5-6_THEMIS', 'PN_dev']\n",
    "}\n",
    "\n",
    "ages_dict = {\n",
    "    \"all_excitatory\"            :   ['1m','3m','6m','10m','1y','2y','4y','ga22','ga24'],\n",
    "    \"all_excitatory_all_ages\"   :   ['1m','3m','6m','10m','1y','2y','4y','6y','10y','16y','20y','40y','ga22','ga24']\n",
    "}\n",
    "\n",
    "output_dir, input_dir, root_dir, tmp_dir, in_dir_from_scenic = set_custom_folders(reference, neurons_set)\n",
    "\n",
    "sel_celltypes  = cells_dict[neurons_set]\n",
    "sel_ages = ages_dict[neurons_set]\n",
    "\n",
    "\n",
    "motif_scan_files = {\n",
    "    \"all_excitatory\"    :           {'L2-3_CUX2': 'L2-3_CUX2.celloracle.parquet',\n",
    "                                    'L4_RORB': 'L4_RORB.celloracle.parquet',\n",
    "                                    'L5-6_THEMIS': 'L5-6_THEMIS.celloracle.parquet',\n",
    "                                    'L5-6_TLE4': '5-6_TLE4.celloracle.parquet',\n",
    "                                    'PN_dev': 'PN_dev.celloracle.parquet'},\n",
    "    \"all_excitatory_all_ages\"    :  {'L2-3_CUX2': 'L2-3_CUX2.celloracle.parquet',\n",
    "                                    'L4_RORB': 'L4_RORB.celloracle.parquet',\n",
    "                                    'L5-6_THEMIS': 'L5-6_THEMIS.celloracle.parquet',\n",
    "                                    'L5-6_TLE4': '5-6_TLE4.celloracle.parquet',\n",
    "                                    'PN_dev': 'PN_dev.celloracle.parquet'}                  \n",
    "}\n",
    "\n",
    "motif_scan_files = motif_scan_files[neurons_set]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a3bc67",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.figsize'] = [6, 4.5]\n",
    "plt.rcParams[\"savefig.dpi\"] = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7d55af8",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(os.path.join(output_dir, 'subseted_rna_andata.h5ad'))\n",
    "print(adata)\n",
    "print(adata.var.shape)\n",
    "print(adata.obs.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a406c63",
   "metadata": {},
   "outputs": [],
   "source": [
    "print([f\"{celltype}: {len(adata.obs.major_clust[adata.obs.major_clust==celltype])}\" for celltype in adata.obs.major_clust.unique()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3e23cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gois = [\"NTRK1\", \"NTRK2\", \"NTRK3\", \"GFRA1\", \"GFRA2\", \"GFRA3\", \"GFRA4\",\n",
    "        \"AHR\", \"ARNT\", \"ARNT2\", \"CLOCK\",\n",
    "        \"AR\",\n",
    "        \"NR1I2\", \"NR1I3\",\n",
    "        \"NR3C1\", \"NR3C2\",\n",
    "        \"ESR1\", \"GPER1\",\n",
    "        \"DIO3\", \"DIO2\",\n",
    "        'RARA', 'ESR2', 'THRB',\n",
    "        \"THRA\", \"THRSP\", \"THRAP3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "889739b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get highly variable genes\n",
    "hvgs = list(adata.var_names[adata.var['highly_variable']])\n",
    "\n",
    "# Filter gois to only include genes present in the dataset\n",
    "gois_present = [gene for gene in gois if gene in adata.var_names]\n",
    "\n",
    "# Combine hvgs and filtered gois, then get unique values\n",
    "combined_genes = hvgs + gois_present\n",
    "unique_genes = pd.Series(combined_genes).unique()\n",
    "\n",
    "# Subset the AnnData object\n",
    "adata = adata[:, unique_genes]\n",
    "\n",
    "print(f\"Number of cells: {adata.n_obs}\")\n",
    "print(f\"Number of genes: {adata.n_vars}\")\n",
    "print(f\"Number of genes of interest found: {len(gois_present)}\")\n",
    "print(f\"Genes of interest not found: {set(gois) - set(gois_present)}\")\n",
    "\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acebaa73",
   "metadata": {},
   "outputs": [],
   "source": [
    "External_data = pd.read_csv(os.path.join(input_dir, \"2023_11_CellOracleProof.tsv\"),delimiter=\"\\t\")\n",
    "External_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8425ce2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "TF_to_TG_dictionary = {}\n",
    "\n",
    "for TF, TGs in zip(External_data.TF, External_data.Target_genes):\n",
    "    # convert target gene to list\n",
    "    TG_list = TGs.replace(\" \", \"\").split(\",\")\n",
    "    # store target gene list in a dictionary\n",
    "    TF_to_TG_dictionary[TF] = TG_list\n",
    "\n",
    "# We invert the dictionary above using a utility function in celloracle.\n",
    "TG_to_TF_dictionary = co.utility.inverse_dictionary(TF_to_TG_dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7404f81c",
   "metadata": {
    "lines_to_next_cell": 1
   },
   "outputs": [],
   "source": [
    "oracles = {}\n",
    "\n",
    "for cell_type, motif_scan_file in motif_scan_files.items():\n",
    "    print(f\"Processing cell type: {cell_type}\")\n",
    "    \n",
    "    # Load the motif scan file for the current cell type\n",
    "    base_GRN = pd.read_parquet(os.path.join(output_dir, motif_scan_file), engine='pyarrow')\n",
    "    \n",
    "    # Create a new CellOracle object for the current cell type\n",
    "    oracle = co.Oracle()\n",
    "    \n",
    "    # Filter the adata object to only include cells from the current cell type\n",
    "    cell_type_adata = adata[adata.obs['major_clust'] == cell_type]\n",
    "    \n",
    "    # Import the filtered adata as raw count\n",
    "    oracle.import_anndata_as_raw_count(cell_type_adata,\n",
    "                                       cluster_column_name=\"major_clust\",\n",
    "                                       embedding_name=\"X_umap\")\n",
    "    \n",
    "    # Import the motif scan data for the current cell type\n",
    "    oracle.import_TF_data(TF_info_matrix=base_GRN)\n",
    "\n",
    "    oracle.addTFinfo_dictionary(TG_to_TF_dictionary)\n",
    "    \n",
    "    # Perform PCA and KNN imputation for the current cell type\n",
    "    oracle.perform_PCA()\n",
    "    n_comps = np.where(np.diff(np.diff(np.cumsum(oracle.pca.explained_variance_ratio_))>0.002))[0][0]\n",
    "    n_comps = min(n_comps, 50)\n",
    "    print(n_comps)\n",
    "\n",
    "    n_cell = oracle.adata.shape[0]\n",
    "    k = int(0.025*n_cell)\n",
    "    print(f\"Auto-selected k is :{k}\")\n",
    "    oracle.knn_imputation(n_pca_dims=n_comps, k=k, balanced=True, b_sight=k*8,\n",
    "                          b_maxl=k*4, n_jobs=n_cpus)\n",
    "    \n",
    "    # Infer the GRN for the current cell type\n",
    "    links = oracle.get_links(cluster_name_for_GRN_unit=\"major_clust\", alpha=10,\n",
    "                             verbose_level=10, n_jobs=n_cpus)\n",
    "    links.filter_links()\n",
    "    links.get_network_score()\n",
    "    \n",
    "    # Save the inferred links for the current cell type\n",
    "    # file_name = os.path.join(output_dir, f\"v1_Herring_links_{cell_type}.celloracle.links\")\n",
    "    # links.to_hdf5(file_path=file_name)\n",
    "\n",
    "    # Store the CellOracle object in the dictionary\n",
    "    oracles[cell_type] = oracle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0abfd133",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the oracles dictionary to a file\n",
    "# with open(os.path.join(output_dir, 'oracles.pkl'), 'wb') as file:\n",
    "#     pickle.dump(oracles, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca82246",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Load the oracles dictionary from the file\n",
    "# with open(os.path.join(output_dir, 'oracles.pkl'), 'rb') as file:\n",
    "#     oracles = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a4ea8fa",
   "metadata": {
    "lines_to_next_cell": 1
   },
   "outputs": [],
   "source": [
    "def plot_arrows_legend(oracle, labels=None, colormap='viridis', scale=1, data_random=False, points_size=5, filename=None):\n",
    "    fig, ax = plt.subplots(figsize=[8, 8])\n",
    "\n",
    "    embedding = oracle.adata.obsm['X_umap']\n",
    "    cluster_labels = oracle.adata.obs[labels]\n",
    "    cluster_categories = pd.Categorical(cluster_labels)\n",
    "    cmap = plt.cm.get_cmap(colormap, len(cluster_categories.categories))\n",
    "\n",
    "    scatter = ax.scatter(embedding[:, 0], embedding[:, 1], c=cluster_categories.codes, cmap=cmap, s=points_size)\n",
    "\n",
    "    # Arrow selection\n",
    "    if data_random:\n",
    "        flow = oracle.flow_rndm\n",
    "    else:\n",
    "        flow = oracle.flow\n",
    "\n",
    "    if hasattr(oracle, \"mass_filter\"):\n",
    "        mass_filter = oracle.mass_filter\n",
    "        gridpoints_coordinates = oracle.flow_grid\n",
    "    else:\n",
    "        mass_filter = np.zeros(flow.shape[0], dtype=bool)\n",
    "        gridpoints_coordinates = embedding\n",
    "\n",
    "    ax.quiver(gridpoints_coordinates[~mass_filter, 0],\n",
    "              gridpoints_coordinates[~mass_filter, 1],\n",
    "              flow[~mass_filter, 0],\n",
    "              flow[~mass_filter, 1],\n",
    "              scale=scale)\n",
    "\n",
    "    ax.axis(\"off\")\n",
    "\n",
    "    if labels is not None:\n",
    "        # Create legend elements based on the cluster categories\n",
    "        legend_elements = [plt.Line2D([0], [0], marker='o', color='w', label=str(label),\n",
    "                                      markerfacecolor=cmap(i), markersize=10)\n",
    "                           for i, label in enumerate(cluster_categories.categories)]\n",
    "        ax.legend(handles=legend_elements, loc='best')\n",
    "\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if filename:\n",
    "        plt.savefig(filename, dpi=300, bbox_inches='tight')\n",
    "    else:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02744869",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize separate CellOracle instances for each cell type\n",
    "cell_type_oracles = {}\n",
    "\n",
    "for cell_type in sel_celltypes:\n",
    "    oracle = oracles[cell_type]\n",
    "    oracle.fit_GRN_for_simulation(alpha=10, use_cluster_specific_TFdict=True)\n",
    "    \n",
    "    cell_type_oracles[cell_type] = oracle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e315d5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_file = True\n",
    "\n",
    "if single_file:\n",
    "    all_sim = []\n",
    "    all_grn = []\n",
    "# Run simulation for each cell type\n",
    "for cell_type, oracle in cell_type_oracles.items():\n",
    "    # Get scores from links\n",
    "    links_file_name = os.path.join(output_dir, f\"v1_Herring_links_{cell_type}.celloracle.links\")\n",
    "    links = co.load_hdf5(links_file_name)\n",
    "    celltype_in_filtered_links = list(links.filtered_links.keys())\n",
    "    \n",
    "    # Double check data integrity\n",
    "    print(f\"celltypes in links.filtered_links: {celltype_in_filtered_links} vs {cell_type} analysed\")\n",
    "    \n",
    "    grn_data = links.filtered_links[celltype_in_filtered_links[0]]\n",
    "\n",
    "    # Run simulation for each GOI\n",
    "    for goi in gois:\n",
    "        try:\n",
    "            if goi in oracle.adata.var_names:\n",
    "                print(f\"{goi} is a regulatory gene in the base GRN for cell type {cell_type}\")\n",
    "\n",
    "                # Plot gene expression\n",
    "                sc.pl.umap(oracle.adata, color=[goi, oracle.cluster_column_name], layer=\"imputed_count\", use_raw=False, cmap=\"viridis\")\n",
    "                plt.savefig(os.path.join(output_dir, f\"gv2_ene_expression_{goi}_{cell_type}.png\"), bbox_inches='tight')\n",
    "                plt.clf()\n",
    "\n",
    "                # Run simulation\n",
    "                oracle.simulate_shift(perturb_condition={goi: 0.0},\n",
    "                                    n_propagation=3)\n",
    "                oracle.estimate_transition_prob(n_neighbors=200,\n",
    "                                                knn_random=True,\n",
    "                                                sampled_fraction=1)\n",
    "                oracle.calculate_embedding_shift(sigma_corr=0.05)\n",
    "\n",
    "                # Plot quiver\n",
    "                fig, ax = plt.subplots(1, 2, figsize=[13, 6])\n",
    "                scale = 50\n",
    "                oracle.plot_quiver(scale=scale, ax=ax[0])\n",
    "                ax[0].set_title(f\"Simulated cell identity shift vector: {goi} KO\")\n",
    "                oracle.plot_quiver_random(scale=scale, ax=ax[1])\n",
    "                ax[1].set_title(f\"Randomized simulation vector\")\n",
    "                plt.savefig(os.path.join(output_dir, f\"v2_quiver_plot_{goi}_{cell_type}.png\"), bbox_inches='tight')\n",
    "                plt.clf()\n",
    "\n",
    "                # Calculate mass and filter\n",
    "                n_grid = 40\n",
    "                oracle.calculate_p_mass(smooth=0.8, n_grid=n_grid, n_neighbors=200)\n",
    "                min_mass = 60\n",
    "                oracle.calculate_mass_filter(min_mass=min_mass, plot=False)\n",
    "                \n",
    "                # Plot simulation flow\n",
    "                fig, ax = plt.subplots(1, 2, figsize=[13, 6])\n",
    "                scale_simulation = 10\n",
    "                oracle.plot_simulation_flow_on_grid(scale=scale_simulation, ax=ax[0])\n",
    "                ax[0].set_title(f\"Simulated cell identity shift vector: {goi} KO\")\n",
    "                oracle.plot_simulation_flow_random_on_grid(scale=scale_simulation, ax=ax[1])\n",
    "                ax[1].set_title(f\"Randomized simulation vector\")\n",
    "                plt.savefig(os.path.join(output_dir, f\"v2_simulation_flow_{goi}_{cell_type}.png\"), bbox_inches='tight')\n",
    "                plt.clf()\n",
    "\n",
    "                # Plot cluster-specific simulation flow\n",
    "                fig, ax = plt.subplots(figsize=[8, 8])\n",
    "                oracle.plot_cluster_whole(ax=ax, s=5)\n",
    "                oracle.plot_simulation_flow_on_grid(scale=scale_simulation, ax=ax, show_background=False)\n",
    "                plt.savefig(os.path.join(output_dir, f\"v2_cluster_specific_simulation_flow_{goi}_{cell_type}.png\"), bbox_inches='tight')\n",
    "                plt.clf()\n",
    "\n",
    "                # Save oracle snapshot for the current GOI and cell type\n",
    "                # file_name = os.path.join(output_dir, f\"v2_Herring_simulation_{goi}_{cell_type}.celloracle.oracle\")\n",
    "                # oracle.to_hdf5(file_name)\n",
    "                \n",
    "                # Evaluate perturbation results\n",
    "                simulated_count = oracle.adata.layers[\"simulated_count\"]\n",
    "                original_count = oracle.adata.X.toarray()\n",
    "                log_fold_change = np.log2(simulated_count + 1) - np.log2(original_count + 1)\n",
    "                \n",
    "                top_n = 1000\n",
    "                cell_types = oracle.adata.obs['major_clust']\n",
    "                gene_names = oracle.adata.var_names\n",
    "                unique_cell_types = cell_types.unique()\n",
    "                print(f\"unique_cell_types: {unique_cell_types}\")\n",
    "                \n",
    "                log_fold_change = log_fold_change.mean(axis=0)\n",
    "                abs_log_fold_change = np.abs(log_fold_change)\n",
    "                sorted_indices = np.argsort(abs_log_fold_change)[::-1]\n",
    "                top_gene_indices = sorted_indices[:top_n]\n",
    "                top_genes = gene_names[top_gene_indices]\n",
    "                top_log_fold_changes = log_fold_change[top_gene_indices]\n",
    "                cell_type_table_data = pd.DataFrame({\n",
    "                    'target': top_genes,\n",
    "                    'log_fold_change': top_log_fold_changes,\n",
    "                })\n",
    "\n",
    "                cell_type_table_data['fold_change'] = np.exp2(cell_type_table_data['log_fold_change'])\n",
    "                \n",
    "                if single_file:\n",
    "                    cell_type_table_data['goi'] = goi\n",
    "                    cell_type_table_data['cell_type'] = cell_type\n",
    "                    all_sim.append(cell_type_table_data)\n",
    "                else:\n",
    "                    cell_type_table_data.to_csv(os.path.join(output_dir, f'v2_scores_sim_{goi}_{cell_type}.csv'), index=False)\n",
    "            else:\n",
    "                print(f\"{goi} is not a regulatory gene in the base GRN for cell type {cell_type}. Skipping simulation.\")      \n",
    "        except ValueError as e:\n",
    "            print(f\"Skipping simulation for gene {goi}: {str(e)}\")\n",
    "\n",
    "        \n",
    "        grn_data = grn_data[grn_data[\"source\"] == goi]\n",
    "        grn_data[\"score\"] = -np.log10(grn_data[\"p\"])\n",
    "        grn_data[\"celltype\"] = cell_type\n",
    "        grn_data = grn_data.rename(columns={\"-logp\": \"X.logp\"})\n",
    "        table_data = grn_data[[\"source\", \"target\", \"coef_mean\", \"coef_abs\", \"p\", \"X.logp\", \"score\", \"celltype\"]]\n",
    "        if single_file:\n",
    "            table_data['goi'] = goi\n",
    "            all_grn.append(table_data)\n",
    "        else:\n",
    "            table_data.to_csv(os.path.join(output_dir, f'v2_scores_grn_{goi}_{cell_type}.csv'), index=False)\n",
    "        \n",
    "# Concatenate the table data from all cell types into a single DataFrame\n",
    "if single_file:\n",
    "    all_sim = pd.concat(all_sim, ignore_index=True)\n",
    "    all_sim.to_csv(os.path.join(output_dir, f'v2_scores_sim_all.csv'), index=False)\n",
    "\n",
    "    # Concatenate all final_table_data DataFrames\n",
    "    all_grn = pd.concat(all_grn, ignore_index=True)\n",
    "    all_grn.to_csv(os.path.join(output_dir, f'v2_scores_grn_all.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
